source/dofs/dof_tools.cc

```diff
@@ -765,6 +765,51 @@ namespace DoFTools
 
 
 
+  template <typename DoFHandlerType>
+  IndexSet
+  extract_dofs_with_support_on(const DoFHandlerType &dof_handler,
+                               const std_cxx11::function< bool(const typename DoFHandlerType::active_cell_iterator &)> &predicate)
+  {
+    IndexSet support_set(dof_handler.n_dofs());
+    // first, get all the dofs which live on the subdomain:
+    std::vector<types::global_dof_index> local_dof_indices;
+    local_dof_indices.reserve (max_dofs_per_cell(dof_handler));
+
+
+    typename DoFHandlerType::active_cell_iterator
+    cell = dof_handler.begin_active(),
+    endc = dof_handler.end();
+    for (; cell!=endc; ++cell)
+      if (predicate(cell))
+        {
+          const unsigned int dofs_per_cell = cell->get_fe().dofs_per_cell;
+          local_dof_indices.resize (dofs_per_cell);
+          cell->get_dof_indices (local_dof_indices);
+          support_set.add_indices(local_dof_indices.begin(),
```

Comment:
other functions in this file first put indices into a `std::set<types::global_dof_index>`, and only in the final step convert the `std::set` to an `IndexSet`. I recall this being the more efficient approach but don't recall the details.